### ISSUE TITLE
rke2 redeployed rke2-ingress-nginx with default config

### ISSUE THREAD

**Issue by dehoej** [rancher/rke2#6668]
**Environmental Info:**
RKE2 Version: 
rke2 version v1.28.11+rke2r1 (6b12d7a783238b72da8450fa1b6ec587cebb79ed)
go version go1.21.11 X:boringcrypto

Node(s) CPU architecture, OS, and Version: 
Linux <hostname> 5.14.0-427.31.1.el9_4.x86_64 #1 SMP PREEMPT_DYNAMIC Fri Aug 9 14:06:03 EDT 2024 x86_64 x86_64 x86_64 GNU/Linux

Cluster Configuration:
3 servers, 8 agents

**Describe the bug:**
I have a custom rke2-ingress-nginx-config.yaml that is very important for our setup. Eg. we set a specific log-format in json and set default ssl-certificate. This has been running fine for a long time, but yesterday suddenly rke-ingress-nginx was redeployed automatically with default settings. I had to make a change in rke2-ingress-nginx-config.yaml to make it redeploy with the right settings. 

**Expected behavior:**
rke2-ingress-nginx will use the provided config in /var/lib/rancher/rke2/server/manifests/rke2-ingress-nginx-config.yaml
**Actual behavior:**
It was redeployed with a default config.

**Additional context / logs:**

`level=error msg="error syncing 'kube-system/rke2-ingress-nginx': handler helm-controller-chart-registration: DesiredSet - Replace Wait batch/v1, Kind=Job kube-system/helm-install-rke2-ingress-nginx for helm-controller-chart-registration kube-system/rke2-ingress-nginx, requeuing"`

**Comment by brandond** [comment]
> I have a custom rke2-ingress-nginx-config.yaml that is very important for our setup

What is in this file? It sounds like you're customizing the ingress-nginx configmap or deployment directly, instead of using a HelmChartConfig resource to customize the chart deployment. If you don't follow the documented process to modify the chart values, any resources managed by the chart will be reverted to the versions generated by the chart.

Ref: https://docs.rke2.io/helm#customizing-packaged-components-with-helmchartconfig

**Comment by dehoej** [comment]
No, I don't edit the config file directly. I edit /var/lib/rancher/rke2/server/manifests/rke2-ingress-nginx-config.yaml. I am following the procedure specified in the documentation exactly: [https://docs.rke2.io/networking/networking_services?_highlight=nginx&_highlight=ingres#nginx-ingress-controller](https://docs.rke2.io/networking/networking_services?_highlight=nginx&_highlight=ingres#nginx-ingress-controller)
By "custom," I just mean that I have edited the config, so it is not the default configuration, and I have followed the documentation.

Let me know if you need any further adjustments!

**Comment by brandond** [comment]
>  I edit /var/lib/rancher/rke2/server/manifests/rke2-ingress-nginx-config.yaml

This isn't a file that RKE2 provides or will create by default. What is in that file? Do you perhaps have something else that is managing its contents?

**Comment by dehoej** [comment]
@brandond 
As stated, I’m simply following the documentation, which specifically mentions that you should use the file: /var/lib/rancher/rke2/server/manifests/rke2-ingress-nginx-config.yaml

![image](https://github.com/user-attachments/assets/c9bcc7f3-96b4-481c-96c8-c95bdfa7af6b)

**Comment by brandond** [comment]
That is an example. You've still not said what content is in that file. If that file contains a HelmChartConfig as shown in the docs example, the values provided via the config should still be honored by the ingress-nginx chart, even if it is redeployed. Please provide the requested information regarding what content you've placed in that file.

**Comment by dehoej** [comment]
It is a HelmChartConfig

Here is the content of the file

```
apiVersion: helm.cattle.io/v1
kind: HelmChartConfig
metadata:
  name: rke2-ingress-nginx
  namespace: kube-system
spec:
  valuesContent: |-
    controller:
      config:
        use-forwarded-headers: "true"
        enable-real-ip: "true"
        log-format-escape-json: "true"
        log-format-upstream: '{ "msec": "$msec", "connection": "$connection", "connection_requests": "$connection_requests", "pid": "$pid", "request_id": "$request_id", "request_length": "$request_length", "remote_addr": "$remote_addr", "remote_user": "$remote_user", "remote_port": "$remote_port", "time_local": "$time_local", "time_iso8601": "$time_iso8601",  "request": "$request", "request_uri": "$request_uri", "args": "$args", "status": "$status", "body_bytes_sent": "$body_bytes_sent", "bytes_sent": "$bytes_sent", "http_referer": "$http_referer", "http_user_agent": "$http_user_agent", "http_x_forwarded_for": "$http_x_forwarded_for", "http_host": "$http_host", "server_name": "$server_name", "request_time": "$request_time", "upstream": "$upstream_addr", "upstream_connect_time": "$upstream_connect_time", "upstream_header_time": "$upstream_header_time", "upstream_response_time": "$upstream_response_time", "upstream_response_length": "$upstream_response_length", "upstream_cache_status": "$upstream_cache_status", "ssl_protocol": "$ssl_protocol", "ssl_cipher": "$ssl_cipher", "scheme": "$scheme", "request_method": "$request_method", "server_protocol": "$server_protocol", "pipe": "$pipe", "gzip_ratio": "$gzip_ratio", "http_cf_ray": "$http_cf_ray"}'
      metrics:
        enabled: true
        serviceMonitor:
          enabled: true
      maxmindLicenseKey: "REDACTED"
      extraArgs:
        default-ssl-certificate: "default/dk-certificate-tls"
        maxmind-edition-ids: "GeoLite2-Country"
        tcp-services-configmap: "kube-system/tcp-services"
      admissionWebhooks:
        patch:
          networkPolicy:
            enabled: true
      containerPort:
        mysql-prod: 13306
        mysql-dev: 53306
      dnsConfig:
        options:
          - name: single-request-reopen
          - name: ndots
            value: "1"

```

And the file is managed as an addon and enabled according to `kubectl get addon -A`.

**Comment by lopf** [comment]
Same here. Did you solve it?

**Comment by dehoej** [comment]
@lopf 
No i did not solve it. I have not experienced the problem more than once.

**Comment by PolarisPT** [comment]
Hi all, 

I seem to have the same issue, but i did not set a HelmChartConfig, and now without any changes my ingress is loading a selfsign certificate instead of the one i stated in rke2-ingress-controller daemonset...
Can you help me out with a HelmChartConfig just to load my custom certificate ?

thanks

**Comment by datton94** [comment]
I'm getting similar issue, I don't want to set Nginx as default ingress class, tried to set `global.systemDefaultIngressClass` empty, null, or the other class but all of them don't help. The Helm job keep installing nginx ingress class with `--set-string global.systemDefaultIngressClass=ingress-nginx`

**Comment by cResults** [comment]
Just experienced the same.  Upgraded rancher which cascaded an update to rke2-ingress-nginx which replaced values content we had added to the rke2-ingress-nginx HelmChartConfig.
@brandond from [your comments here](https://github.com/rancher/rke2/issues/6668#issuecomment-2329694959), and the [documentation found here](https://docs.rke2.io/networking/networking_services?_highlight=nginx&_highlight=ingres#nginx-ingress-controller), it seems that we should be able to make edits to the rke2-ingress-nginx HelmChartConfig valuesContent and expect them to remain through updates.  That is currently not the case.  
We were updating from rancher 2.8.x to 2.10.2.

**Comment by brandond** [comment]
> @cResults 
> an update to rke2-ingress-nginx which replaced values content we had added to the rke2-ingress-nginx HelmChartConfig.

There are no HelmChartConfig resources bundled with RKE2. There is nothing in RKE2 that will alter the HelmChartConfig.  Did you perhaps make some changes to the HelmChartConfig resource in the cluster, but forgot to update a manifest that you had placed on disk?

If you're using the manifests dir to deploy resources, you must ensure that those stay up to date with any changes you make by hand. Preferably you just wouldn't edit things in the cluster at all, but rather make you changes to the files on disk and let the deploy controller update the cluster.

**Comment by cResults** [comment]
Wow!! Thank you for your rapid response!

Based upon what you are saying, I'm currently suspecting that the rancher update made the change.  We only added the following to rke2-ingress-nginx HelmChartConfig
```yaml
    tcp:
      11122: "prod/our-service:11122"
```
to expose a service via tcp.  This had been working for months.
Then Friday we updated rancher and this tcp setting was removed.

We are moving towards using Argo CD to enforce that this doesn't get changed unless its change in GitHub.  We would appreciate any thoughts you have for or against this idea.

**Comment by brandond** [comment]
You'd really need to look at how you created that HelmChartConfig, and where you're making changes to it.

Did you create it via Rancher Additional Manifest? Did you place the file in the manifests folder on disk? Did you use `kubectl` to create/apply it by hand?

Whatever method you used to create it, should also be how you update it.

**Comment by cResults** [comment]
@brandond Thank you!  I didn't set the original values as I inherited the current cluster.  The only value I set was the tcp property mentioned above and you were correct I set it directly on the HelmChartConfig manifest via the Rancher UI.  
Your answer pointed me to the "Additional Manifest" field in the Cluster UI.  I wasn't aware of that field and that is definitely where the rke2-ingress-nginx HelmChartConfig was originally set.  Thank you!
